File: autoconfigure-adapter/src/main/java/org/springframework/boot/autoconfigure/mongo/embedded/EmbeddedMongoInitializer.java
/*
 * Copyright 2002-2018 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.boot.autoconfigure.mongo.embedded;

import java.io.IOException;

import de.flapdoodle.embed.mongo.MongodExecutable;

import de.flapdoodle.embed.mongo.config.MongodConfig;
import de.flapdoodle.embed.process.config.RuntimeConfig;
import org.springframework.boot.autoconfigure.mongo.MongoProperties;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.support.GenericApplicationContext;

public class EmbeddedMongoInitializer implements ApplicationContextInitializer<GenericApplicationContext> {

	private final MongoProperties properties;

	private final EmbeddedMongoProperties embeddedProperties;

	public EmbeddedMongoInitializer(MongoProperties mongoProperties, EmbeddedMongoProperties embeddedProperties) {
		this.properties = mongoProperties;
		this.embeddedProperties = embeddedProperties;
	}

	@Override
	public void initialize(GenericApplicationContext context) {
		// , this.embeddedProperties
		EmbeddedMongoAutoConfiguration configuration = new EmbeddedMongoAutoConfiguration(this.properties);
		context.registerBean(MongodConfig.class, () -> {
			try {
				return configuration.embeddedMongoConfiguration(this.embeddedProperties);
			}
			catch (IOException e) {
				e.printStackTrace();
			}
			return null;
		});
		context.registerBean("embeddedMongoServer", MongodExecutable.class, () ->
				configuration.embeddedMongoServer(context.getBean(MongodConfig.class), context.getBean(RuntimeConfig.class), context), definition -> {
			definition.setInitMethodName("start");
			definition.setDestroyMethodName("stop");
		});
		context.registerBean(RuntimeConfig.class, () -> new EmbeddedMongoAutoConfiguration.RuntimeConfigConfiguration().embeddedMongoRuntimeConfig(context.getBeanProvider(DownloadConfigBuilderCustomizer.class)));
	}
}

File: kofu/src/main/kotlin/org/springframework/boot/autoconfigure/mongo/embedded/EmbeddedMongoInitializer.kt
package org.springframework.boot.autoconfigure.mongo.embedded

import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.beans.factory.config.BeanDefinitionCustomizer
import org.springframework.beans.factory.getBean
import org.springframework.beans.factory.support.AbstractBeanDefinition
import org.springframework.boot.autoconfigure.mongo.MongoProperties
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.support.GenericApplicationContext
import org.springframework.context.support.registerBean

internal class EmbeddedMongoInitializer(private val mongoProperties: MongoProperties, private val embeddedMongoProperties: EmbeddedMongoProperties) : ApplicationContextInitializer<GenericApplicationContext> {

	override fun initialize(context: GenericApplicationContext) {
		val config = EmbeddedMongoAutoConfiguration.RuntimeConfigConfiguration().embeddedMongoRuntimeConfig()
		val configuration = EmbeddedMongoAutoConfiguration(mongoProperties, embeddedMongoProperties, context, config)
		context.registerBean { configuration.embeddedMongoConfiguration() }

		// TODO See with Juergen to avoid this ugly AbstractBeanDefinition cast
		val customizer = BeanDefinitionCustomizer { bd: BeanDefinition ->
			val beanDefinition = bd as AbstractBeanDefinition
			beanDefinition.initMethodName = "start"
			beanDefinition.destroyMethodName = "stop"
		}
		context.registerBean("embeddedMongoServer", customizer) { configuration.embeddedMongoServer(context.getBean())}
		context.registerBean { config }
	}


}